### ISSUE TITLE
heartbeat problem on non_blocking consumers

### ISSUE THREAD

**Issue by decrypted** [php-amqplib/php-amqplib#508]
Oki so i debugged on this for like 5+ hours and are still a little dizzy. Yet i wanna open the report now, hoping it will be good enough to parse.

Disclaimer: As i´m new php-amqplib internals, it could be on the completely wrong foot here, unlikely as of the deep level debugging yet i only did find close but not exact issues reported.

So the setup is like https://github.com/php-amqplib/php-amqplib/blob/master/demo/amqp_consumer_non_blocking.php.

But enabling heartbeat. Lets say 10 seconds. So after 10 seconds the *server* does send an heartbeat package. The socket gets an signal and does enter the wait.

https://github.com/php-amqplib/php-amqplib/blob/5bc0b01d9498a4c4d5df047b18076d3be99597d3/demo/amqp_consumer_non_blocking.php#L89

BTW: this will be also the first time check_heartbeat will run. While debugging i added very dirty reflection invokes to call the protected getIo and check_heartbeat in regular intervals - as in this setup the missing reads do cause no check_heartbeat. Yet thats a side story i more or less hope.

The problem now is that the *wait* will (potentially) never return, as the heartbeat gets eaten by this https://github.com/php-amqplib/php-amqplib/blob/c2144bae989befe90c376395bc46b2ee5bc6fa7d/PhpAmqpLib/Connection/AbstractConnection.php#L571. This then causes the non blocking design all of the sudden to become very blocking and eventually the server just also killing the connections.

I dont really have a good suggestion how to really fix or change this - ideas are very welcome. IMHO the main challenge is, that all calling functions of that AbstractConnection function do wait for command packages and check for that type rightfully so. Which means just handling the heartbeat in a more or less simple class override or higher up is not 

trivial. My current workaround does raise an exception on server induced heartbeat and have a specific handler around the - expected to be - non blocking wait to silence it.

Happy to help!

**Comment by andrefigueira** [comment]
@decrypted so If I'm understanding correctly, you're getting into situations where the connection has been dropped by the server silently, so you end up waiting for messages that will never come?

**Comment by decrypted** [comment]
@andrefigueira i wanna run in non_blocking mode. Yet the lib is blocking when there is a new package from the server and i request processing it gets stuck if there only is a heartbeat message. The lib is getting the heartbeat message, does eat it and wait for the the next message in a blocking way.

I dont know how to workaround this as of the structure of the lib, only option i have seen so far is raising an exception on a heartbeat message.

**Comment by boryashkin** [comment]
There is another blocking spot:
https://github.com/php-amqplib/php-amqplib/blob/0f90b3d8bc50403458f0eefbcba7d1e2329dd0f6/PhpAmqpLib/Wire/IO/StreamIO.php#L231

**Comment by decrypted** [comment]
as of finding out about https://github.com/php-amqplib/php-amqplib/pull/520 i did look into this again
here is my solution for now, where i dont need to fork


```php
class AMQPStreamConnectionHeartbeat_Exception extends Exception {};

class AMQPStreamConnectionHeartbeat extends PhpAmqpLib\Connection\AMQPStreamConnection {

	/** @var bool If set to true, raise on heartbeat as of https://github.com/php-amqplib/php-amqplib/issues/508 */
	static public $raise_on_heartbeat = true;
	
	/**
	 * Waits for a frame from the server destined for a particular channel.
	 *
	 * @param string $channel_id
	 * @param int $timeout
	 * @return array
	 */
	protected function wait_channel($channel_id, $timeout = 0)
	{
		// Keeping the original timeout unchanged.
		$_timeout = $timeout;
		while (true) {
			$now = time();
			try {
				list($frame_type, $frame_channel, $payload) = $this->wait_frame($_timeout);
			}
			catch ( AMQPTimeoutException $e ) {
				if ( $this->heartbeat && microtime(true) - ($this->heartbeat*2) > $this->last_frame ) {
					$this->debug->debug_msg("missed server heartbeat (at threshold * 2)");
					$this->setIsConnected(false);
					throw new AMQPRuntimeException("Missed server heartbeat");
				}

				throw $e;
			}

			$this->last_frame = microtime(true);

			if ($frame_channel === 0 && $frame_type === 8) {
				// skip heartbeat frames and reduce the timeout by the time passed
				$this->debug->debug_msg("received server heartbeat");
				if($_timeout > 0) {
					$_timeout -= time() - $now;
					if($_timeout <= 0) {
						// If timeout has been reached, throw the exception without calling wait_frame
						throw new AMQPTimeoutException("Timeout waiting on channel");
					}
				}

				if (self::$raise_on_heartbeat)
					throw new AMQPStreamConnectionHeartbeat_Exception('heartbeat');

				continue;

			} else {

				if ($frame_channel == $channel_id) {
					return array($frame_type, $payload);
				}

				// Not the channel we were looking for.  Queue this frame
				//for later, when the other channel is looking for frames.
				// Make sure the channel still exists, it could have been
				// closed by a previous Exception.
				if (isset($this->channels[$frame_channel])) {
					array_push($this->channels[$frame_channel]->frame_queue, array($frame_type, $payload));
				}

				// If we just queued up a method for channel 0 (the Connection
				// itself) it's probably a close method in reaction to some
				// error, so deal with it right away.
				if (($frame_type == 1) && ($frame_channel == 0)) {
					$this->wait();
				}
			}
		}
	}
} 
```

and then 
```php
// This is non-blocking, since there is data ready on the socket for this channel.
try {
	$item->channel->wait();
} catch (AMQPStreamConnectionHeartbeat_Exception $e) {
	//hack to handle Heartbeat ok
};		
```

**Comment by lukebakken** [comment]
Please test these scenarios with `2.9.0-rc1`, which will be released this week. Thank you.

**Comment by lukebakken** [comment]
Assuming resolved by `2.9.0`. If not, please comment on this issue.

**Comment by jason-engage** [comment]
Hi, I upgraded to 2.9.1 and I seem to be getting a 'Missed Server Heartbeat' Error.

I was on 2.8.1 and I was getting a 'Connection timeout after 3 sec error', so I upgraded. I think that went away but now the new heartbeat.

I've been running this script for over a year with no errors at all. The 3sec error just started a week or so ago.  Perhaps I have more traffic than I used to? And its causing some glitch? Maybe I'm generally doing something wrong in my channel/queue declaration?

**Comment by lukebakken** [comment]
@jason-engage Connection timeouts are indicative of network issues. More information is necessary to help out. Please open a new issue in this repo with at least this information:

* AMQP broker and broker version
* Php version
* Description of how script is run (via web server, standalone process)
* Complete script (ideally) or version that demonstrates general method flow
* How long does processing a message take on average before an ack is sent back?
* AMQP broker logs and application logs
